File: internal/server/transition.go
package server

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"

	"churn/internal/domain"
	"churn/internal/event"
	"churn/internal/writer"
)

// transitionRequest is the body of POST /things/{id}/transition.
//
// Propose: {"state": "st_…"} — if the target state has active semantics
// (and the thing is not already active), the response is a PROPOSAL, not a
// commit: the concrete allocation set (domain.ProposeAllocations) plus an
// opaque token binding thing, state, assignment, and the projection seq it
// was computed at.
//
// Confirm: {"state": "st_…", "confirm": true, "proposal": "<token>"} —
// submits transition + allocations as ONE batch. The writer re-validates in
// its critical section; if the world drifted (another actor took the last
// unit), the commit returns 409 with details.fresh_proposal (a new token, or
// null when nothing is feasible now) so the client can re-confirm.
//
// Transitions to non-active states commit directly; leaving active
// semantics closes all open allocations in the same batch. Active→active
// transitions commit directly with allocations untouched.
type transitionRequest struct {
	State    string `json:"state"`
	Confirm  bool   `json:"confirm,omitempty"`
	Proposal string `json:"proposal,omitempty"`
}

// proposalClaim is the decoded content of a proposal token. The token is
// stateless: base64(JSON) of exactly these fields — the server keeps no
// proposal registry, and the writer's critical-section re-validation (not
// the token's seq) is the drift guard. BasedOnSeq is disclosure, not a
// precondition: a confirm against a moved-on-but-still-feasible world
// commits fine (§5: conflicts are for drift that invalidates, not for any
// drift).
type proposalClaim struct {
	Thing       string          `json:"thing"`
	State       string          `json:"state"`
	BasedOnSeq  int64           `json:"based_on_seq"`
	Allocations []allocationRow `json:"allocations"`
}

// allocationRow is one line of a proposed assignment.
type allocationRow struct {
	Requirement string `json:"requirement"`
	Resource    string `json:"resource"`
	Quantity    int    `json:"quantity"`
}

// proposalDTO is the wire form of a proposal response.
type proposalDTO struct {
	Token       string          `json:"token"`
	Thing       string          `json:"thing"`
	State       string          `json:"state"`
	BasedOnSeq  int64           `json:"based_on_seq"`
	Allocations []allocationRow `json:"allocations"`
}

func encodeProposal(c proposalClaim) (proposalDTO, error) {
	b, err := json.Marshal(c)
	if err != nil {
		return proposalDTO{}, err
	}
	return proposalDTO{
		Token: base64.RawURLEncoding.EncodeToString(b),
		Thing: c.Thing, State: c.State, BasedOnSeq: c.BasedOnSeq,
		Allocations: c.Allocations,
	}, nil
}

func decodeProposal(token string) (proposalClaim, *apiError) {
	b, err := base64.RawURLEncoding.DecodeString(token)
	if err != nil {
		return proposalClaim{}, errBadRequest("proposal token is not valid base64")
	}
	var c proposalClaim
	if err := json.Unmarshal(b, &c); err != nil {
		return proposalClaim{}, errBadRequest("proposal token does not decode")
	}
	return c, nil
}

func toRows(props []domain.ProposedAllocation) []allocationRow {
	rows := make([]allocationRow, len(props))
	for i, pr := range props {
		rows[i] = allocationRow{Requirement: pr.Requirement, Resource: pr.Resource, Quantity: pr.Quantity}
	}
	return rows
}

// transitionResult is the response of a committed transition or re-propose.
type transitionResult struct {
	Committed bool   `json:"committed"`
	Thing     string `json:"thing"`
	State     string `json:"state,omitempty"`
	Seq       int64  `json:"seq,omitempty"`
	Batch     string `json:"batch,omitempty"`
	// Proposal is set (with Committed false) on the propose leg.
	Proposal *proposalDTO `json:"proposal,omitempty"`
	// Opened / Closed list the allocation ids the batch opened and closed.
	Opened []string `json:"opened,omitempty"`
	Closed []string `json:"closed,omitempty"`
}

// postTransition implements POST /api/v1/things/{id}/transition (§2.5,
// §5.1): propose→confirm into active states, direct commit otherwise.
func (s *Server) postTransition(rw http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	var req transitionRequest
	if e := decodeJSON(r, &req); e != nil {
		writeError(rw, e)
		return
	}
	p := s.w.Projection()
	th, ok := p.Things[id]
	if !ok {
		writeError(rw, errNotFound(id))
		return
	}
	if len(th.Children) > 0 {
		writeError(rw, &apiError{status: http.StatusUnprocessableEntity,
			kind: domain.KindCompositeState, ids: []string{id},
			message: "thing " + id + " is a composite: its state is a computed rollup, not a fact"})
		return
	}
	st, ok := p.States[req.State]
	if !ok {
		writeError(rw, &apiError{status: http.StatusUnprocessableEntity,
			kind: domain.KindUndefinedReference, ids: []string{req.State},
			message: "state " + req.State + " is not defined"})
		return
	}
	expected, e := expectedFromIfMatch(r, id)
	if e != nil {
		writeError(rw, e)
		return
	}

	curActive := len(th.Children) == 0 && p.SemanticOf(th) == event.SemanticActive
	targetActive := st.Semantic == event.SemanticActive

	if targetActive && !curActive && !req.Confirm {
		// Propose leg: compute the concrete assignment, commit nothing.
		props, feasible := p.ProposeAllocations(id)
		if !feasible {
			writeError(rw, &apiError{status: http.StatusConflict,
				kind: domain.KindInfeasible, ids: []string{id},
				message: "no feasible allocation of the thing's requirements onto free units exists right now"})
			return
		}
		dto, err := encodeProposal(proposalClaim{
			Thing: id, State: req.State, BasedOnSeq: p.LastSeq, Allocations: toRows(props),
		})
		if err != nil {
			writeError(rw, mapError(err))
			return
		}
		writeJSON(rw, http.StatusOK, transitionResult{Committed: false, Thing: id, State: req.State, Proposal: &dto})
		return
	}

	var cmds []writer.Command
	var opened, closed []string
	if targetActive && !curActive {
		// Confirm leg: transition + the proposal's allocations as ONE batch.
		claim, e := decodeProposal(req.Proposal)
		if e != nil {
			writeError(rw, e)
			return
		}
		if claim.Thing != id || claim.State != req.State {
			writeError(rw, errBadRequest("proposal token binds thing %s state %s, request names thing %s state %s",
				claim.Thing, claim.State, id, req.State))
			return
		}
		cmds = append(cmds, writer.Command{
			Type: event.TypeThingStateChanged, V: 1, Entity: id,
			Payload: &event.ThingStateChanged{State: req.State},
		})
		for _, row := range claim.Allocations {
			// The rows come from the CLIENT-HELD token: validate their shape
			// like any other request payload, so a tampered token is a 400,
			// never an internal writer error.
			pl := &event.AllocationOpened{
				Thing: id, Resource: row.Resource, Quantity: row.Quantity, Requirement: row.Requirement,
			}
			if e := validatePayload(pl); e != nil {
				writeError(rw, e)
				return
			}
			alID, e := s.mint(event.PrefixAllocation)
			if e != nil {
				writeError(rw, e)
				return
			}
			opened = append(opened, alID)
			cmds = append(cmds, writer.Command{
				Type: event.TypeAllocationOpened, V: 1, Entity: alID, Payload: pl,
			})
		}
	} else {
		// Direct commit: non-active target (closing all open allocations in
		// the same batch when leaving active semantics), or active→active
		// with allocations untouched.
		if curActive && !targetActive {
			for _, alID := range p.OpenAllocationsOf(id) {
				pl := &event.AllocationClosed{}
				if e := validatePayload(pl); e != nil { // symmetry with the open set
					writeError(rw, e)
					return
				}
				closed = append(closed, alID)
				cmds = append(cmds, writer.Command{
					Type: event.TypeAllocationClosed, V: 1, Entity: alID, Payload: pl,
				})
			}
		}
		cmds = append(cmds, writer.Command{
			Type: event.TypeThingStateChanged, V: 1, Entity: id,
			Payload: &event.ThingStateChanged{State: req.State},
		})
	}

	evs, err := s.w.Submit(s.actor, cmds, expected)
	if err != nil {
		s.writeTransitionConflict(rw, err, id, req.State)
		return
	}
	writeJSON(rw, http.StatusOK, transitionResult{
		Committed: true, Thing: id, State: req.State,
		Seq: evs[len(evs)-1].Seq, Batch: evs[0].Batch,
		Opened: opened, Closed: closed,
	})
}

// writeTransitionConflict maps a rejected transition/confirm batch. For the
// drift kinds — capacity, infeasible_allocation, allocation_coverage (a
// requirement changed between propose and confirm), stale_version — it
// answers 409 and attaches details.fresh_proposal: a NEW proposal computed
// against the current projection (null when nothing is feasible), so the
// client can re-confirm directly.
func (s *Server) writeTransitionConflict(rw http.ResponseWriter, err error, thing, state string) {
	ae := mapError(err)
	var de *domain.Error
	if errors.As(err, &de) {
		switch de.Kind {
		case domain.KindCapacity, domain.KindInfeasible, domain.KindAllocationCoverage, domain.KindStaleVersion:
			ae.status = http.StatusConflict
			ae.details = map[string]any{"fresh_proposal": nil}
			cur := s.w.Projection()
			if props, feasible := cur.ProposeAllocations(thing); feasible {
				if dto, encErr := encodeProposal(proposalClaim{
					Thing: thing, State: state, BasedOnSeq: cur.LastSeq, Allocations: toRows(props),
				}); encErr == nil {
					ae.details["fresh_proposal"] = dto
				}
			}
		}
	}
	writeError(rw, ae)
}

// postRepropose implements POST /api/v1/things/{id}/repropose: the §2.5
// one-click reconciliation for the allocations-out-of-step badge — ONE batch
// closes the thing's open allocations and opens a fresh feasible assignment
// of its CURRENT requirements, while the thing stays active throughout. The
// assignment is computed server-side at request time; the writer's
// critical-section re-validation guards the remaining window, so a drifted
// re-propose returns a conflict and the client simply retries.
func (s *Server) postRepropose(rw http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	if r.Body != nil && r.ContentLength > 0 {
		var empty struct{}
		if e := decodeJSON(r, &empty); e != nil {
			writeError(rw, e)
			return
		}
	}
	p := s.w.Projection()
	th, ok := p.Things[id]
	if !ok {
		writeError(rw, errNotFound(id))
		return
	}
	if len(th.Children) > 0 || p.SemanticOf(th) != event.SemanticActive {
		writeError(rw, &apiError{status: http.StatusUnprocessableEntity,
			kind: domain.KindAllocation, ids: []string{id},
			message: "re-propose applies to an active leaf thing"})
		return
	}
	expected, e := expectedFromIfMatch(r, id)
	if e != nil {
		writeError(rw, e)
		return
	}

	props, feasible := p.ReproposeAllocations(id)
	if !feasible {
		writeError(rw, &apiError{status: http.StatusConflict,
			kind: domain.KindInfeasible, ids: []string{id},
			message: "no feasible replacement assignment exists for the thing's current requirements"})
		return
	}

	var cmds []writer.Command
	var opened, closed []string
	for _, alID := range p.OpenAllocationsOf(id) {
		closed = append(closed, alID)
		cmds = append(cmds, writer.Command{
			Type: event.TypeAllocationClosed, V: 1, Entity: alID,
			Payload: &event.AllocationClosed{},
		})
	}
	for _, pr := range props {
		alID, e := s.mint(event.PrefixAllocation)
		if e != nil {
			writeError(rw, e)
			return
		}
		opened = append(opened, alID)
		cmds = append(cmds, writer.Command{
			Type: event.TypeAllocationOpened, V: 1, Entity: alID,
			Payload: &event.AllocationOpened{
				Thing: id, Resource: pr.Resource, Quantity: pr.Quantity, Requirement: pr.Requirement,
			},
		})
	}
	if len(cmds) == 0 {
		// No open allocations and no requirements: nothing to reconcile.
		writeJSON(rw, http.StatusOK, transitionResult{Committed: false, Thing: id})
		return
	}
	evs, err := s.w.Submit(s.actor, cmds, expected)
	if err != nil {
		writeError(rw, mapError(err))
		return
	}
	writeJSON(rw, http.StatusOK, transitionResult{
		Committed: true, Thing: id,
		Seq: evs[len(evs)-1].Seq, Batch: evs[0].Batch,
		Opened: opened, Closed: closed,
	})
}

File: web/src/ui/transition.ts
// ui/transition.ts — state transitions with the §2.5 propose→confirm flow.

import { api, ApiError, Proposal, Semantic, StateDef, Thing } from '../api';
import { h, select } from '../dom';
import { closeModal, openModal } from '../modal';
import { store } from '../store';
import { showError, toast } from '../toast';
import { reqText } from './bits';

/** semanticActions maps a thing's current situation to offered transitions. */
export function actionsFor(t: Thing): { label: string; semantic: Semantic; kind?: string }[] {
  if (t.composite) return [];
  const sem = store.semanticOf(t);
  switch (sem) {
    case undefined: // never started: statutes blocked/ready/resource_blocked
    case 'pending':
      return [{ label: 'Start', semantic: 'active' }];
    case 'active':
      return [
        { label: 'Finish', semantic: 'satisfied' },
        { label: 'Pause', semantic: 'paused' },
        { label: 'Abandon', semantic: 'abandoned' },
      ];
    case 'paused':
      return [
        { label: 'Resume', semantic: 'active' },
        { label: 'Abandon', semantic: 'abandoned' },
      ];
    case 'satisfied':
    case 'abandoned':
      return [{ label: 'Reopen', semantic: 'pending' }];
    default:
      return [];
  }
}

/** transitionTo runs the full flow: pick a state of the semantic (if several
 * are defined), then commit — via the proposal modal when entering active. */
export async function transitionTo(t: Thing, semantic: Semantic): Promise<void> {
  const states = store.statesBySemantic(semantic);
  if (states.length === 0) {
    toast(`No state with semantic "${semantic}" is defined — add one in the vocabulary.`, 'error');
    return;
  }
  let state = states[0]!;
  if (states.length > 1) {
    const picked = await pickState(semantic, states);
    if (!picked) return;
    state = picked;
  }
  await commitTransition(t, state);
}

function pickState(semantic: Semantic, states: StateDef[]): Promise<StateDef | null> {
  return new Promise((resolve) => {
    const sel = select(states.map((s) => ({ value: s.id, label: s.name })), states[0]!.id);
    const body = h('div', null,
      h('p', { class: 'muted' }, `Several states share the "${semantic}" semantic — pick the one to record.`),
      sel,
      h('div', { class: 'modal-actions' },
        h('button', { class: 'btn', onclick: () => { closeModal(); resolve(null); } }, 'Cancel'),
        h('button', {
          class: 'btn btn-primary',
          onclick: () => {
            const s = states.find((x) => x.id === sel.value) ?? null;
            closeModal();
            resolve(s);
          },
        }, 'Choose')));
    openModal('Choose state', body);
  });
}

async function commitTransition(t: Thing, state: StateDef): Promise<void> {
  const curSem = store.semanticOf(t);
  const entersActive = state.semantic === 'active' && curSem !== 'active';
  try {
    if (!entersActive) {
      if (curSem === 'satisfied' && state.semantic === 'pending') {
        toast(`Reopening ${t.name} — recorded as an ordinary transition.`, 'info');
      }
      const res = await api.transition(t.id, { state: state.id });
      if (res.committed) toast(`${t.name} → ${state.name}`, 'ok', 2500);
      await store.refresh();
      return;
    }
    // Propose leg.
    const res = await api.transition(t.id, { state: state.id });
    if (res.committed) { await store.refresh(); return; }
    if (res.proposal) showProposalModal(t, state, res.proposal);
  } catch (e) {
    showError(e);
  }
}

/** showProposalModal renders the assignment table; Confirm commits transition
 * + allocations as one batch. On 409 drift the fresh proposal replaces the
 * stale one with a notice (details.fresh_proposal). */
export function showProposalModal(t: Thing, state: StateDef, proposal: Proposal, notice?: string): void {
  const rows = proposal.allocations;
  const body = h('div', null,
    notice ? h('div', { class: 'notice notice-warn' }, notice) : null,
    h('p', null,
      'Starting ', h('b', null, t.name), ' as ', h('b', null, state.name),
      '. Proposed assignment (based on seq ', String(proposal.based_on_seq), '):'),
    rows.length === 0
      ? h('p', { class: 'muted' }, 'No requirements — nothing to allocate.')
      : h('table', { class: 'table' },
        h('thead', null, h('tr', null,
          h('th', null, 'Requirement'), h('th', null, 'Resource'), h('th', null, 'Units'))),
        h('tbody', null, ...rows.map((r) => {
          const req = store.requirements.find((x) => x.id === r.requirement);
          return h('tr', null,
            h('td', null, req ? reqText(req.quantity, req.capabilities, req.resource) : r.requirement),
            h('td', null, store.name(r.resource)),
            h('td', null, String(r.quantity)));
        }))),
    h('div', { class: 'modal-actions' },
      h('button', { class: 'btn', onclick: closeModal }, 'Cancel'),
      h('button', {
        class: 'btn btn-primary',
        onclick: async () => {
          try {
            const res = await api.transition(t.id, { state: state.id, confirm: true, proposal: proposal.token });
            if (res.committed) {
              closeModal();
              toast(`${t.name} started (${res.opened?.length ?? 0} allocation(s) opened)`, 'ok');
              await store.refresh();
            }
          } catch (e) {
            if (e instanceof ApiError && 'fresh_proposal' in e.details) {
              const fresh = e.details['fresh_proposal'] as Proposal | null;
              if (fresh) {
                showProposalModal(t, state, fresh,
                  'The world drifted since this proposal was computed — review the fresh assignment below.');
              } else {
                closeModal();
                toast('The world drifted and no feasible assignment exists right now.', 'error', 8000);
              }
              void store.refresh();
              return;
            }
            showError(e);
          }
        },
      }, 'Confirm assignment')));
  openModal('Allocation proposal', body);
}

/** repropose runs the §2.5 one-click atomic reconciliation. */
export async function repropose(t: Thing): Promise<void> {
  try {
    const res = await api.repropose(t.id);
    if (res.committed) {
      toast(`Re-proposed: closed ${res.closed?.length ?? 0}, opened ${res.opened?.length ?? 0} allocation(s).`, 'ok');
    } else {
      toast('Nothing to reconcile.', 'info');
    }
    await store.refresh();
  } catch (e) {
    showError(e);
  }
}
